### DISCUSSION TITLE
Camera

### DISCUSSION THREAD

**Discussion by rhettg** [rhettg/stubborn#8] [Ideas]
Ok, for my next trick, I'm going to try to integrate a camera in Stubborn.

As discussed in [this twitter thread](https://twitter.com/rhettford/status/1411089198354878464) I've been experimenting with the dubious idea of a camera on an Arduino. This is a challenge because:

* An Arduino is pretty slow, images can be a lot of data.
* There is no USB or other standard peripheral interface. Everyone doing vision stuff using a Pi (or other).
* Stubborn is out of pins.
* The radio link is low-bandwidth.

But options do exist. My first experiment with the [minature TTL Serial camera from Adafruit](https://www.adafruit.com/product/1386) I'm calling a failure.

![image](https://user-images.githubusercontent.com/50074/137603936-9310168d-5a65-430a-b0b7-7d388a8a6c6d.png)

But I'm having more luck with the [ArduCam Mini-2MP-Plus](https://www.arducam.com/docs/spi-cameras-for-arduino/hardware/arducam-shield-mini-2mp-plus/).

![out](https://user-images.githubusercontent.com/50074/137603972-f214a281-1ea1-4d2f-b245-3532b5484ac8.jpg)

My test rig though has little to do with how Stubborn is wired. This is on a Arduino Nano 33-BLE:

![317632C6-CEBF-4FA8-BA2F-7E633240DADB_1_105_c](https://user-images.githubusercontent.com/50074/137603991-07ec01f1-218f-4cd1-abef-e35cac597594.jpeg)

Now I'm going to attempt to integrate this into Stubborn somehow and chronicle this effort here.

**Comment by rhettg** [comment]
This ArduCam is interesting because it uses TWO busses, SPI + I2C. I don't really understand why, but beyond it being annoying it almost seems like it should be able to fit with the existing Stubborn layout.

The RFM69 Transceiver is already using SPI so I need to bypass my existing wiring so another device can be on the same bus lines: MOSI, MISO and SCK. On Stubborn that's pins 12, 11 and 13. Like some sort of bomb surgeon I've hacked into it with a mess of wires and jumpers:

<img width="1001" alt="image" src="https://user-images.githubusercontent.com/50074/137604496-8a2e722d-789e-4f2b-8f7d-fdb2172a16fc.png">

If this works I'll need to solder into my breakout board.

**Comment by rhettg** [comment]
I want to make sure before spending a bunch of time integrating the camera into my rover software that this thing is actually going to work.

I rebuilt the Camera demo software which is the hacked up demo code enhanced to write base64 image data to Serial for the Arduino Uno. (See https://gist.github.com/rhettg/4795ed7131aff2876e069338ff00eab3)

I'm using [Platform IO](https://platformio.org) for this, which I don't have a ton of experience for. I wasn't really sure how to build for a new environment. Turns out I just need to specify the environment in my platformio.ini file:

```!ini
[env:uno]
platform = atmelavr
framework = arduino
board = uno
```

It Built! It Uploaded! But then.... my terminal just showed funny characters. Turns out I need just one more setting:

```
monitor_speed = 115200
```

Because the Uno defaults to 9600 but I had told it to run faster.

Ok, so now we're communicating, but it still failed:

    SPI interface Error!

After some poking around I looked up the Uno reference and it turns out I somehow reversed MOSI and MISO when I audited the pins I was using. Swap the blue and green wire and we have:

```
SPI interface OK.
Checking chip
OV2640 detected.
Snap in 3 secs...
Started capture
Capture done
Storing 7176 byte image.
...
wrote 6658 bytes
done
```

🎉

**Comment by rhettg** [comment]
Took a little more than an hour but I'm almost remembering how to use my own Rover framework. I got some new events wired up, some telemetry and even added it to the gds:

<img width="842" alt="image" src="https://user-images.githubusercontent.com/50074/137840537-09fa4a5f-59bc-42ad-bd77-5be73b621b5f.png">

The most involved part of the change is probably in https://github.com/rhettg/stubborn/pull/9/commits/8878da6a8bebf0cf71d0483759d7c6090a9a5c05 where I introduce a `SNAP` command.

This command:

1. Starts the camera capture.
2. Kicks off a timer to poll for completion.
3. On completion, records how many bytes were saved.

The next step should be to start pulling data off and see just how long it will take me to download my little 4000 byte image. However, I'm seeing some communication instability already just in getting a good response from the command. I think I need to iterate on COM a little.

**Comment by tilnow** [comment]
is using video ascii an option? less bandwidth no?

**Reply by rhettg** [reply]
That's a cool idea. I probably couldn't encode video ascii, or even jpg ascii on the Uno with my limited memory, but that would be an interesting fallback option for the next iteration 🤔

**Comment by rhettg** [comment]
Well this is encouraging:

![cam](https://user-images.githubusercontent.com/50074/138207545-50461a21-e727-4ad8-9050-a590b950f841.jpg)

This is, for the record, the camera wired up to the rover. After issuing the `SNAP` command, the data is streamed back over the radio and written to disk on the ground station. I'm pretty impressed.

But there are two problems, one of them is really obvious:

1. The image is kinda messed up
2. The transfer is pretty unreliable

These might be related, however I am carefully counting the packets sent and packets received. For this image they are supposed to have matched, so I'm not entirely sure what's wrong. I wonder actually if there is noise on the SPI connection on the Arduino itself. As a reminder, this is still just plugged into a breadboard.

One of the most important pieces of data I wanted to establish was just how fast I could send an image over this rfm69 transceiver. Sending single packet commands back and forth doesn't exactly push the limits. What I've found is that with all my debugging on, it downloads in about 30 seconds. What's extra cool is that I can see the image buffer decrement in the telemetry.

The surprising thing was when I disabled debug logging on the rover (which displays a bunch of stuff on the Serial line), the image was sent in under a second, fast enough to never see telemetry. *BUT* it lost half the packets.

The data stream implementation was the simplest possible thing that just might work... and it almost did. Without having any space on the rover to buffer data (the Uno only has 2k total and I'm down to having about 700 free) I'm limited in how much I can buffer and resend. It's probably preferable to slow the entire thing down and do a a request-reply cycle for each packet of data. I think that will only slow down the send to about 30 seconds and the reliability will be worth it.

**Comment by rhettg** [comment]
After much bug fixing, I have a working camera:
![cam caprica](https://user-images.githubusercontent.com/50074/140868956-e447d177-6014-454d-8095-2e78edfd7d1a.jpg)

The req-reply methodology works and seems to be very resilient. It is certainly slow though. I have some tuning I think I can do that will hopefully improve throughput. In the meantime, can wire this up to the rover?

If you recall, I seem to be short one pin on the Arduino so I was (re)using one of the motor control pins. To re-integrate the rover, I need my pins back. 

My first attempt was to try to use pin 0/1 which are used for the UART / Serial port. This does not work. At least I can't make it work. I created some `#define` options to build the rover software with serial disabled (no `Serial.begin`). I also tried shortcutting even including the serial.h header file at all:

```
#ifndef ENABLE_SERIAL
// Skip importing the Serial library if we're not using it.
#define HardwareSerial_h
#endif
```

No dice. 

After more googling I came across [a tantalizing discussion thread](https://forum.arduino.cc/t/using-digital-pin-0-and-1-just-like-any-other-digital-pin/462591/7) where someone was trying to figure out the same thing. The conversation ended with a simple throwaway comment:

> You know that you can use the analog pins as digital pins too, right?

![image](https://user-images.githubusercontent.com/50074/141901202-c91db178-157a-404d-a4c7-0eb878ef9649.gif)

Ok, so I have plenty of pins. This Rover is GO! 🎉

**Comment by rhettg** [comment]
Forgot to post this, but I did a demo of Stubborn (with camera) at a recent Yak Rover meeting:

https://www.youtube.com/watch?v=41EsEE77_lc
